File: swara-music/src/main/java/com/swara/music/readers/JavaReader.java
package com.swara.music.readers;

import java.io.InputStream;
import java.io.ObjectInputStream;

import com.swara.music.MusicReader;
import com.swara.music.elements.Song;

/**
 * Reads a {@link Song} from a stream encoded using the standard Java Serialization API. This reader
 * is included as a way of benchmarking the performance of alternative techniques.
 */
public final class JavaReader implements MusicReader {

    @Override
    @SuppressWarnings({ "unchecked" })
    public Song read(InputStream input) throws Exception {
        final ObjectInputStream reader = new ObjectInputStream(input);
        return (Song) reader.readObject();
    }

}

File: swara-music/src/main/scala/com/swara/music/readers/JavaReader.scala
package com.swara.music
package readers

import java.io.{InputStream, ObjectInputStream}
import scala.util.Try

/**
 * Reads a [[Song]] from a stream encoded using the standard Java Serialization Api. This reader is
 * included as a way of benchmarking the performance of alternative deserialization techniques.
 * Because each [[MusicElement]] is necessarily [[Serializable]], the JavaReader is guaranteed to
 * work on all data structures.
 */
object JavaReader extends MusicReader {

  override def read(in: InputStream): Try[Song] =
    Try(new ObjectInputStream(in).readObject().asInstanceOf[Song])

}
